Client/Telemetry/UnifiedTelemetry.swift

```diff
@@ -68,22 +68,36 @@ class UnifiedTelemetry {
             return outputDict
         }
 
-       Telemetry.default.beforeSerializePing(pingType: MobileEventPingBuilder.PingType) { (inputDict) -> [String: Any?] in
-           var outputDict = inputDict
+        Telemetry.default.beforeSerializePing(pingType: MobileEventPingBuilder.PingType) { (inputDict) -> [String: Any?] in
+            var outputDict = inputDict
 
-           var settings: [String: String?] = inputDict["settings"] as? [String: String?] ?? [:]
+            var settings: [String: String?] = inputDict["settings"] as? [String: String?] ?? [:]
 
-           let searchEngines = SearchEngines(prefs: profile.prefs, files: profile.files)
-           settings["defaultSearchEngine"] = searchEngines.defaultEngine.engineID ?? "custom"
+            let searchEngines = SearchEngines(prefs: profile.prefs, files: profile.files)
+            settings["defaultSearchEngine"] = searchEngines.defaultEngine.engineID ?? "custom"
 
-           if let windowBounds = UIApplication.shared.keyWindow?.bounds {
-               settings["windowWidth"] = String(describing: windowBounds.width)
-               settings["windowHeight"] = String(describing: windowBounds.height)
-           }
+            if let windowBounds = UIApplication.shared.keyWindow?.bounds {
+                settings["windowWidth"] = String(describing: windowBounds.width)
+                settings["windowHeight"] = String(describing: windowBounds.height)
+            }
+
+            outputDict["settings"] = settings
 
-           outputDict["settings"] = settings
-           return outputDict
-       }
+            // App Extension telemetry requires reading events stored in prefs, then clearing them from prefs.
+            if let extensionEvents = profile.prefs.arrayForKey(PrefsKeys.AppExtensionTelemetryEventArray) as? [[String: String]],
+                var pingEvents = outputDict["events"] as? [[Any?]] {
+                profile.prefs.removeObjectForKey(PrefsKeys.AppExtensionTelemetryEventArray)
+
+                extensionEvents.forEach { extensionEvent in
+                    let category = UnifiedTelemetry.EventCategory.appExtensionAction.rawValue
+                    let newEvent = TelemetryEvent(category: category, method: extensionEvent["method"] ?? "", object: extensionEvent["object"] ?? "")
+                    pingEvents.append(newEvent.toArray())
+                }
+                outputDict["events"] = pingEvents
```

Comment:
Isn't this going to overwrite existing events? Should probably be appending to `outputDict["events"]` array instead of replacing it.

Reply:
These nested ummutable data structures make this code read a bit weird. 
On line 88, the ` var pingEvents = outputDict["events"]` grabs a mutable copy of the events. Then pingEvents gets appended to, then assigned back to  outputDict["events"].
That is, `outputDict["events"].append(..)` won't compile due to immutability. I can assign to it however.